### ISSUE TITLE
postscript

### ISSUE THREAD

**Issue by KG-NINJA** [KG-NINJA/SuzumeBachiBlowdart#27]
#### Change 1: Add TimeSeriesSplit function

Add this new function before the main tuning logic:
```python
from sklearn.model_selection import TimeSeriesSplit
import pandas as pd

def create_temporal_splits(df: pd.DataFrame, 
                          train_ratio: float = 0.7,
                          val_ratio: float = 0.15) -> tuple:
    """
    Create time-series aware train/val/test splits.
    
    CRITICAL: This prevents data leakage by ensuring no future data 
    is used during training.
    """
    # Ensure data is sorted by date
    df = df.sort_values('Date').reset_index(drop=True)
    
    n = len(df)
    train_end = int(n * train_ratio)
    val_end = int(n * (train_ratio + val_ratio))
    
    train_data = df.iloc[:train_end].copy()
    val_data = df.iloc[train_end:val_end].copy()
    test_data = df.iloc[val_end:].copy()
    
    print(f"  Train: {len(train_data)} samples | {train_data['Date'].min()} to {train_data['Date'].max()}")
    print(f"  Val:   {len(val_data)} samples | {val_data['Date'].min()} to {val_data['Date'].max()}")
    print(f"  Test:  {len(test_data)} samples | {test_data['Date'].min()} to {test_data['Date'].max()}")
    
    # Validation: ensure no overlap
    assert train_data['Date'].max() < val_data['Date'].min(), "Train/Val overlap detected!"
    assert val_data['Date'].max() < test_data['Date'].min(), "Val/Test overlap detected!"
    
    return train_data, val_data, test_data
```

#### Change 2: Update hyperparameter grid

Replace the existing `param_grid` with:
```python
param_grid = {
    # Increase complexity slightly
    'max_depth': [4, 5, 6],              # Was: [3]
    'learning_rate': [0.05, 0.1, 0.15],  # Was: [0.01]
    
    # Add regularization (KEY CHANGE)
    'reg_alpha': [0.1, 1, 10],           # L1 regularization - NEW
    'reg_lambda': [1, 10, 100],          # L2 regularization - NEW
    'min_child_weight': [3, 5, 7],       # Prevents overfitting - NEW
    'gamma': [0, 0.1, 0.5],              # Min loss reduction - NEW
    
    # Keep existing
    'subsample': [0.7, 0.8],
    'colsample_bytree': [0.7, 0.8]
}
```

#### Change 3: Add early stopping

In the model training section, update to:
```python
from xgboost import XGBClassifier

model = XGBClassifier(
    **best_params,
    objective='binary:logistic',
    eval_metric='logloss',
    early_stopping_rounds=50,  # NEW - stops if no improvement
    random_state=42
)

# Fit with validation set for early stopping
model.fit(
    X_train, y_train,
    eval_set=[(X_train, y_train), (X_val, y_val)],  # Monitor both
    verbose=False
)
```

#### Change 4: Report all metrics separately

Update the results dictionary to include all three datasets:
```python
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Get predictions for all three sets
train_pred = model.predict(X_train)
val_pred = model.predict(X_val)
test_pred = model.predict(X_test)

results = {
    'ticker': ticker,
    'best_params': best_params,
    
    # TRAIN metrics
    'train_accuracy': accuracy_score(y_train, train_pred),
    'train_precision': precision_score(y_train, train_pred),
    'train_recall': recall_score(y_train, train_pred),
    
    # VALIDATION metrics
    'val_accuracy': accuracy_score(y_val, val_pred),
    'val_precision': precision_score(y_val, val_pred),
    'val_recall': recall_score(y_val, val_pred),
    
    # TEST metrics (most important)
    'test_accuracy': accuracy_score(y_test, test_pred),
    'test_precision': precision_score(y_test, test_pred),
    'test_recall': recall_score(y_test, test_pred),
    
    # Overfitting indicator
    'overfitting_gap': accuracy_score(y_train, train_pred) - accuracy_score(y_test, test_pred)
}

print(f"  Train Acc: {results['train_accuracy']:.4f}")
print(f"  Val Acc:   {results['val_accuracy']:.4f}")
print(f"  Test Acc:  {results['test_accuracy']:.4f}")
print(f"  Gap:       {results['overfitting_gap']:.4f}")

# Warning if overfitting is high
if results['overfitting_gap'] > 0.15:
    print(f"  ⚠️  WARNING: High overfitting detected!")
```

#### Change 5: Update main tuning function

Locate the main function (likely `tune_xgboost_for_ticker` or similar) and replace the data splitting logic:
```python
def tune_xgboost_for_ticker(ticker: str, df: pd.DataFrame) -> dict:
    """Tune XGBoost with proper time-series validation"""
    
    print(f"\n{'='*70}")
    print(f"Tuning XGBoost for {ticker}")
    print(f"{'='*70}")
    
    # OLD (REMOVE):
    # X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
    
    # NEW (ADD):
    train_data, val_data, test_data = create_temporal_splits(df)
    
    X_train = train_data.drop(['Date', 'Target'], axis=1)
    y_train = train_data['Target']
    X_val = val_data.drop(['Date', 'Target'], axis=1)
    y_val = val_data['Target']
    X_test = test_data.drop(['Date', 'Target'], axis=1)
    y_test = test_data['Target']
    
    # Continue with grid search...
```

---

### Testing Instructions

After implementing, test with one ticker first:
```bash
# Test with single ticker
python -c "
from scripts.phase2_xgboost_tuning import tune_xgboost_for_ticker
import yfinance as yf

df = yf.download('AAPL', period='2y')
# Add your feature engineering here
result = tune_xgboost_for_ticker('AAPL', df)
print(f'Test Accuracy: {result[\"test_accuracy\"]:.2%}')
print(f'Overfitting Gap: {result[\"overfitting_gap\"]:.2%}')
"
```

Expected output should show:
- Test accuracy: 60-70% (not 100%)
- Overfitting gap: < 15%

---

@codex Please implement these changes exactly as specified. The key goals are:
1. Use temporal splits (no data leakage)
2. Add regularization parameters
3. Implement early stopping
4. Report train/val/test separately
